### ISSUE TITLE
[source-quickbooks] token expires every 24h

### ISSUE THREAD

**Issue by berz01** [airbytehq/airbyte#49842]
### Connector Name

Quickbooks

### Connector Version

QuickBooks v3.0.26

### What step the error happened?

None

### Relevant information

Strange behavior on disconnects every 24 hours with the QuickBooks connector.

It doesn't appear the token refresh every 24 hours like its supposed to any more? My oauth production apps are totally fine. Is anyone else experiencing only running after the initial 24 hours of a successful test & entering a new refresh token?

![Image](https://github.com/user-attachments/assets/3c67924a-423e-4fc9-850a-cfded7e5f612)

Let me know if I can assist with updating the refresh capabilities

### Relevant log output

```shell
Configuration check failed
'Unauthorized. Please ensure you are authenticated correctly.'
```

### Contribute

- [x] Yes, I want to contribute

**Comment by natikgadzhi** [comment]
I see. Must be that the token can't refresh quickly enough. @bnchrch @erohmensing @pabloescoder do any of our new OAuth things put us in a good spot to test them on Quickbooks tomorrow morning, see if we can fix the obvious problems?

**Comment by pabloescoder** [comment]
@natikgadzhi @bnchrch @erohmensing I think I know what's wrong. 
I read the QB docs, they invalidate the refresh token every time it's used to request a new access token & send a new refresh token. 

So we'll have to use the "override config with refresh token response" checkbox in the builder which creates a refresh_token_updater in the yaml.

I'll raise a PR for this. 

![Screenshot_2024_1218_102113.png](https://github.com/user-attachments/assets/fda85660-5a0c-45d0-9c70-286328f5d1f2)

**Comment by berz01** [comment]
thanks guys, I knew something was up and I know how important QuickBooks is to your marketplace and customers. I could tell it wasn't refreshing with the 'newest' refresh token. 

Let me know if you need more testing you can dm me directly

**Comment by berz01** [comment]
Any luck here @natikgadzhi ? We had to pause our airbyte project until QB is fixed, which I'm hoping is a simple adjustment to the refresh function

**Comment by pabloescoder** [comment]
Hey @berz01 
We just released a new version of QB (v4.0.0) with the fix. 
It's a major version upgrade because the credentials object was removed from the config and all the config fields were moved to the root level so you might need to repopulate the config fields. 
TLDR: QB v4.0.0 solves this issue!

**Comment by berz01** [comment]
Holy shit good work Pablo! i'll check try now.

**Comment by berz01** [comment]
I'm still running into issues. It appears there is still a bug that does not use the refresh token and only uses the current saved access token. The saved access token is only valid for 60 minutes and the refresh token is 101 days.

@pabloescoder 

![Image](https://github.com/user-attachments/assets/d1053ab6-c4b8-43cc-b285-ba369ee76730)

**Comment by pabloescoder** [comment]
@berz01 In QB, whenever we use the refresh token to request for a new access token (after the previous one expires), it returns not only a new access token, but also a new refresh token (so both the refresh and access tokens need to be overwritten, which is what v4.0.0 is doing currently) 

The sync rate for QB has improved a lot and it is working for other users. Can you please confirm if you're using v4.0.0? 

If you're on v4.0.0 and you're still facing the problem, could you please elaborate a bit more on what error you are seeing?

**Comment by berz01** [comment]
Hello heres a link to the logs: https://cloud.airbyte.com/workspaces/608abb32-c6c3-42c2-bdbe-2a5b1417b0d6/connections/a9309a4f-8de2-4986-afba-aaad85152700/timeline?openLogs=true&jobId=28782115

We're using 4.0.1

Whenever we make the connection, it works. Then it fails the next 24 hours because its failing to refresh or retrieve a new access token it seems.

**Comment by benjamin-smith** [comment]
I am still seeing this issue on both v4.0.0 and today on v4.0.4. I tried updating my old connection, as well as creating a new one. It works for 1 hour, but the refresh token still doesn't seem to work / do anything and I need to reauthenticate to QBO. Is there something I am missing here?

**Comment by fredericogg** [comment]
Same issue here. I'm using the QuickBooks connector v4.0.4 on Airbyte Cloud. The first sync works great, but after 24 hours, the sync fails because the refresh token has changed. I guess Airbyte is trying to use an old token.

**Comment by benjamin-smith** [comment]
> I am still seeing this issue on both v4.0.0 and today on v4.0.4. I tried updating my old connection, as well as creating a new one. It works for 1 hour, but the refresh token still doesn't seem to work / do anything and I need to reauthenticate to QBO. Is there something I am missing here?

Sorry, false report on my end. I had the "Token Expiry Date" field set to when the refresh token expired (3 months out), not the access token (1 hour out). I updated this and it has been running every hour as intended.
